log/term/colorlogger.go

```diff
@@ -0,0 +1,149 @@
+package term
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"sync"
+
+	"github.com/go-kit/kit/log"
+)
+
+// Color is the abstract color, the zero value is the Default.
+type Color uint8
+
+const (
+	NoColor = Color(iota)
+	Red
+	Green
+	Yellow
+	Blue
+	Magenta
+	Cyan
+	White
+	Default
+
+	maxColor
+)
+
+var resetColorBytes = []byte("\x1b[0m")
+var fgColorBytes [][]byte
+var bgColorBytes [][]byte
+
+func init() {
+	for color := NoColor; color < maxColor; color++ {
+		fgColorBytes = append(fgColorBytes, []byte(fmt.Sprintf("\x1b[%dm", 30+color)))
+		bgColorBytes = append(bgColorBytes, []byte(fmt.Sprintf("\x1b[%dm", 40+color)))
+	}
+}
+
+type FgBgColor struct {
+	Fg, Bg Color
+}
+
+func (c FgBgColor) IsZero() bool {
+	return c.Fg == NoColor && c.Bg == NoColor
+}
+
+// NewColorLogger returns a log.Logger which writes colored logs to w. It
+// colors whole records based on the FgBgColor returned by the color function.
+// Log events are formatted by the Logger returned by newLogger.
+func NewColorLogger(w io.Writer, newLogger func(io.Writer) log.Logger, color func(keyvals ...interface{}) FgBgColor) log.Logger {
+	if color == nil {
+		panic("color func nil")
+	}
```

Comment:
Hmm... I'm afraid I'm not convinced. I find an explicit test for IsTty at the callsite to be too user-hostile. The semantics of ColorLogger are: wrap another logger, and colorize the output according to some declared rules, if possible. If the final Writer can't handle color codes, there's no reason to punish me for the mistake of not properly sanitizing my operating context, as that detection •can• happen automatically. Put another way: colorization is a pure value-add, and should fail gracefully; a broken ColorLogger should no-op, rather than produce mangled output.

If colorizing output on Windows isn't possible without wiring in mattn's lib, we should ship with that package, too, same as Simon's.